src/Microsoft.ML.Data/StaticPipe/TrainerEstimatorReconciler.cs

```diff
@@ -0,0 +1,335 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Utilities;
+
+namespace Microsoft.ML.Data.StaticPipe.Runtime
+{
+    /// <summary>
+    /// General purpose reconciler for a typical case with trainers, where they accept some generally
+    /// fixed number of inputs, and produce some outputs where the names of the outputs are fixed.
+    /// Authors of components that want to produce columns can subclass this directly, or use one of the
+    /// common nested subclasses.
+    /// </summary>
+    public abstract class TrainerEstimatorReconciler : EstimatorReconciler
+    {
+        private readonly PipelineColumn[] _inputs;
+        private readonly string[] _outputNames;
+
+        /// <summary>
+        /// The output columns. Note that subclasses should return exactly the same items each time,
+        /// and the items should correspond to the output names passed into the constructor.
+        /// </summary>
+        protected abstract IEnumerable<PipelineColumn> Outputs { get; }
+
+        /// <summary>
+        /// Constructor for the base class.
+        /// </summary>
+        /// <param name="inputs">The set of inputs</param>
+        /// <param name="outputNames">The names of the outputs, which we assume cannot be changed</param>
```

Comment:
real question: do we put a period in the sentences of the <summary>section, but not in the sentences in <params>?

Reply:
Having complete sentences in the summary is essential. About params, I'm less sure about that, but I think I prefer to have them, I'll try to add them.

---
In reply to: [215322337](https://github.com/dotnet/machinelearning/pull/837#discussion_r215322337) [](ancestors = 215322337)